### PR TITLE
Unify drag & drop handling.

### PR DESCRIPTION
**Pull request by cyrill-wyss**
Back port drag & drop handling changes from scout 10.0.x:
- If42e98749929f203f2b54b741a9f82e0e7f93cac
- I5690860699b6865b7d9f34e6063b8d7abeac6584
- Ia118153649bd69c2c9746401ecba528b0ae5ab0e

301777
Signed-off-by: Cyrill Wyss <cyrill.wyss@bsi-software.com>

### CODE REVIEW

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/table/Table.js [hunk 1]
```diff
@@ -4534,12 +4538,31 @@ scout.Table.prototype._installDragAndDropHandler = function(event) {
...
+  }
+};
+
+
+scout.Table.prototype._installDragAndDropHandler = function() {
+  if (this.dragAndDropHandler) {
+    return;
+  }
+  this.dragAndDropHandler = this._createDragAndDropHandler();
+  if (!this.dragAndDropHandler) {
```
Comment (awegmueller) [inline org.eclipse.scout.rt.ui.html/src/main/js/scout/table/Table.js:4558]: I know, this code was backported: but in which case should _createDragAndDropHandler return null? I found no implementation that returns null (or a JS-doc which explains in which case this should be used).

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/table/Table.js [hunk 2]
```diff
@@ -4534,12 +4538,31 @@ scout.Table.prototype._installDragAndDropHandler = function(event) {
...
+};
+
+scout.Table.prototype._installOrUninstallDragAndDropHandler = function() {
+  if (this.dropType && this.enabledComputed) {
+    this._installDragAndDropHandler();
+  } else {
+    this._uninstallDragAndDropHandler();
+  }
+};
+
```
Comment (awegmueller) [inline org.eclipse.scout.rt.ui.html/src/main/js/scout/table/Table.js:4552]: Unnecessary blank line

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/form/fields/clipboardfield/ClipboardField.js [hunk 3]
```diff
@@ -11,8 +11,8 @@
 scout.ClipboardField = function() {
   scout.ClipboardField.parent.call(this);
 
-  this.dropType = 0;
-  this.dropMaximumSize = scout.dragAndDrop.DEFAULT_DROP_MAXIMUM_SIZE;
+  this.dropType = null;
+  this.dropMaximumSize = null;
```
Comment (awegmueller) [inline org.eclipse.scout.rt.ui.html/src/main/js/scout/form/fields/clipboardfield/ClipboardField.js]: This variable is never read in this file. Still required?

Comment (awegmueller) [reply org.eclipse.scout.rt.ui.html/src/main/js/scout/form/fields/clipboardfield/ClipboardField.js]: Is it really required to call _renderDropType in _renderProperties? This should be done by the super class anyway. Maybe check all sub-classes of FormField for unnecessary _render* method calls related to D&D

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/filechooser/FileInput.js [hunk 4]
```diff
@@ -120,6 +118,52 @@ scout.FileInput.prototype.setText = function(text) {
...
+    dropType: function() {
+      return scout.dragAndDrop.SCOUT_TYPES.FILE_TRANSFER;
+    },
+    dropMaximumSize: function() {
+      return this.maximumUploadSize;
+    }.bind(this)
+  });
+};
+
+scout.FileInput.prototype._installOrUninstallDragAndDropHandler = function() {
```
Comment (awegmueller) [inline org.eclipse.scout.rt.ui.html/src/main/js/scout/filechooser/FileInput.js:140]: The same code for the 3 functions below exist in other classes. Maybe move to a common helper class?

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/util/DragAndDropHandler.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+scout.DragAndDropHandler.prototype._onDrop = function(event) {
+  if (this.supportedScoutTypes.indexOf(scout.dragAndDrop.SCOUT_TYPES.FILE_TRANSFER) >= 0 &&
+    (this.dropType() & scout.dragAndDrop.SCOUT_TYPES.FILE_TRANSFER) === scout.dragAndDrop.SCOUT_TYPES.FILE_TRANSFER && // NOSONAR
+    scout.dragAndDrop.dataTransferTypesContainsScoutTypes(event.originalEvent.dataTransfer, scout.dragAndDrop.SCOUT_TYPES.FILE_TRANSFER)) {
+    if (!this.onDrop ||
+      !(event.originalEvent.dataTransfer.files instanceof FileList)) {
+      return;
+    }
+    var files = Array.from(event.originalEvent.dataTransfer.files);
```
Comment (awegmueller) [inline org.eclipse.scout.rt.ui.html/src/main/js/scout/util/DragAndDropHandler.js]: IE doesn't know Array.from

https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/from

Not a good idea to be used in Scout 9 (no Babel). Use utility functions from Scout arrays, or simply myArray.slice()

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/util/DragAndDropHandler.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    try {
+      this.validateFiles(files);
+    } catch (error) {
+      this._validationFailed(files, error);
+      return;
+    }
+    event.stopPropagation();
+    event.preventDefault();
+    var formattedDropEvent = {
```
Comment (awegmueller) [inline org.eclipse.scout.rt.ui.html/src/main/js/scout/util/DragAndDropHandler.js]: Why is this event "formatted"? I'd simply inline this object into the onDrop call

File: org.eclipse.scout.rt.ui.html/src/main/js/scout/table/Table.js (hunk 1 above)
Comment (cyrill-wyss) [reply org.eclipse.scout.rt.ui.html/src/main/js/scout/table/Table.js:4558]: It allows for an extension to completely remove drag & drop handling by overriding _createDragAndDropHandler and returning null.

**Review by awegmueller — APPROVED** [review]

**Pull request merged**